src/phone-input/stateful-phone-input-container.js

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright (c) 2018-2019 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+import React from 'react';
+import {countries, STATE_CHANGE_TYPE} from './constants.js';
+
+import type {
+  StatefulPhoneInputContainerPropsT,
+  StateT,
+  StateReducerT,
+  StateChangeT,
+  StateChangePayloadT,
+  CountryChangeEventT,
+} from './types.js';
+
+const getCountry = iso => {
+  return countries.find(({id}) => id === iso) || {dialCode: ''};
+};
+
+const defaultStateReducer: StateReducerT = (type, nextState) => nextState;
+
+export default class StatefulPhoneInputContainer extends React.Component<
+  StatefulPhoneInputContainerPropsT,
+  StateT,
+> {
+  static defaultProps = {
+    initialState: {
+      inputValue: `+${getCountry('US').dialCode} `,
+      countryValue: getCountry('US'),
+    },
+    overrides: {
+      Input: {},
+      CountrySelect: {},
+      CountrySelectDropdown: {},
+      CountrySelectDropdownListItem: {},
+      CountrySelectDropdownFlagColumn: {},
+      CountrySelectDropdownNameColumn: {},
+      CountrySelectDropdownDialcodeColumn: {},
+    },
+    onInputChange: () => {},
+    onCountryChange: () => {},
+    stateReducer: defaultStateReducer,
+  };
+
+  state = {...this.props.initialState};
+
+  internalSetState = (type: StateChangeT, payload: StateChangePayloadT) => {
+    const nextState: StateT = this.internalStateReducer(type, payload);
+    this.setState(this.props.stateReducer(type, nextState));
+  };
+
+  internalStateReducer = (type: StateChangeT, payload: StateChangePayloadT) => {
+    switch (type) {
+      case STATE_CHANGE_TYPE.inputValueChange:
+        if (typeof payload === 'string') {
+          return {...this.state, inputValue: payload};
+        } else {
+          return this.state;
+        }
+      case STATE_CHANGE_TYPE.countryValueChange: {
+        if (typeof payload !== 'string') {
+          // Replace (if possible) the current country dialcode
+          let newInputValue = this.state.inputValue.replace(
+            `+${this.state.countryValue.dialCode}`,
+            `+${payload.dialCode}`,
+          );
+          // If the replacement did nothing, just use the new dialcode
+          newInputValue =
+            this.state.inputValue === newInputValue
+              ? `+${payload.dialCode} `
```

Comment:
Will it result in dropping the phone number entered? Also, this says that the input value won't consistently contain a country code, which is not clear if intended.

Reply:
i can update this so that if we cannot replace a recognizable dial code in the input string, we prepend it to the current input value instead.

so if the input value is `4882399` and you select `US` from dropdown it will prepend the country code: `+1 4882399`. (right now it would change the string to `+1 `)

Regarding the second point, i think this touches on the awkwardness around my decision to have the country code inserted into the input string, rather than being a separate un-editable part of the UI. It allows the user to delete the country code from the input.

originally, i had the state and presentation separate. i changed it because i wanted to let the user opt out of entering a country code, say if for some reason our dropdown did not cover their needs.

but now i think maybe i should have the state and presentation stay separate. in the case where a developer did not want to require the country code, i can add a prop `requireCountryCode` which when set to `false`, adds an `Other` option to the dropdown.

even then, why use this component if no country code is required? you would be better off just using your own text field and adding a mask if you did not require any country code logic.